sys/picolibc_syscalls_default/syscalls.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (C) 2019 Koen Zandberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @defgroup    sys_picolibc PicoLibc system call
+ * @ingroup     sys
+ * @brief       PicoLibc system call
+ * @{
+ *
+ * @file
+ * @brief       PicoLibc system call implementations
+ *
+ * @author      Koen Zandberg <koen@bergzand.net>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <stdio.h>
+
+#include "log.h"
+#include "periph/pm.h"
+#include "stdio_base.h"
+
+uint8_t sbrk[1];
+
+/**
+ * @brief Exit a program without cleaning up files
+ *
+ * If your system doesn't provide this, it is best to avoid linking with subroutines that
+ * require it (exit, system).
+ *
+ * @param n     the exit code, 0 for all OK, >0 for not OK
+ */
+void _exit(int n)
+{
+    LOG_INFO("#! exit %i: powering off\n", n);
+    pm_off();
+    while(1);
+}
+
+/**
+ * @brief Send a signal to a thread
+ *
+ * @param[in] pid the pid to send to
+ * @param[in] sig the signal to send
+ *
+ * @return    always returns -1 to signal error
+ */
+__attribute__ ((weak))
+int kill(pid_t pid, int sig)
+{
+    (void) pid;
+    (void) sig;
+    errno = ESRCH;                         /* not implemented yet */
+    return -1;
+}
+
+/* File descriptor pointers */
+int picolibc_get(FILE *file)
+{
+    (void)file;
+    char c = 0;
+    stdio_read(&c, 1);
+    return c;
+}
+
+int picolibc_put(char c, FILE *file)
+{
+    (void)file;
+    stdio_write(&c, 1);
+    return 1;
+}
+FILE picolibc_stdio =
+    FDEV_SETUP_STREAM(picolibc_put, picolibc_get, NULL, _FDEV_SETUP_RW);
```

Comment:
Will there be a flush function in case of CDC ACM stdio?

Reply:
Yes, what's the API for that? I didn't see one in the newlib code.

Reply:
Ok, I reviewed the CDC ACM implementation and that doesn't have any buffering internally; each call to stdio_write causes a separate USB transaction. So, the current code will 'work', although output will be slow. I've added a buffer for output, and also a pile of additional picolibc bits at

 https://github.com/keith-packard/RIOT/tree/pr/libc/picolibc_initial

Reply:
Want to push that pile of additional picolibc bits again? :D 
Or do you want to pull them @bergzand ?

I kind of missed this comment here, I see you hooked up VFS too now.
For the buffering, I wonder if CDC ACM should handle this, but that can be changed later.

Reply:
I've rebased my changes on top of @bergzand's current branch and pushed them to my branch again.

https://github.com/keith-packard/RIOT/tree/pr/libc/picolibc_initial

I also added a config option so that by default the integer-only printf/scanf code is use, but the full float code can be selected.

Reply:
As for the CDC ACM buffering, it would be nice to include that in the underlying implementation, but that would require exposing a flush function that all users called as needed.

Reply:
> I've rebased my changes on top of @bergzand's current branch and pushed them to my branch again.

And pulled them in.

Reply:
I note that you left out the stdout buffering code. That makes USB performance pretty poor; it might be nice to add that back in and plan on removing it if/when the underlying ACM code adds buffering. I'm not really sure it should though -- you'd have to modify all users of that code to know about the buffering so that they could flush when needed.